### ISSUE TITLE
可视化问题

### ISSUE THREAD

**Issue by gaowenjie-star** [Sense-X/Co-DETR#136]
您好，很感谢您的工作，我使用了您的可视化代码，但是我有一个问题关于multi-scale features，据我所知，deformable detr，在encoder的操作之前，会将backbone出来的特征将H,W进行合并，然后将4维的多维特征进行铺平，然后再进行后续的操作得到feature，您这里的multi-scale feature的维度是LxBxCxHxW，请问您是将后续得到的的feature进行逆操作得到的吗，还是说我的理解有问题，期待您的回复。

**Comment by TempleX98** [comment]
把encoder输出的不同尺度的feature抽出来，插值到最大特征的维度再concat

**Comment by gaowenjie-star** [comment]
> 把encoder输出的不同尺度的feature抽出来，插值到最大特征的维度再concat

感谢您的回复，您是否可以分享对应的代码呢

**Comment by TempleX98** [comment]
https://github.com/Sense-X/Co-DETR/blob/d01ef8b657cc3711c59759cf863fe5ab9aa50e76/projects/models/co_deformable_detr_head.py#L160
这里是从encoder输出中抽出不同scale特征的代码，后面把每个scale的特征尺寸都插值到`mlvl_feats[0]`的尺寸就可以

**Comment by JiaPiangLi** [comment]
我在训练过程中在通道维度做l2-norm之后，特征图所有信息都变得很接近，热力图也没有明显峰值（第一个epoch），这会随着训练进行而改善吗，还是说这是我的计算过程有误。

**Comment by JiaPiangLi** [comment]
for i in range(level_start_index.shape[0]):
                # 取出单阶段（分辨率）尺度特征
                stage_memory = memory[:, level_start_index[i]:index[i+1], :]
                # 恢复H, W维度, 缩放到统一尺寸
                stage_memory = stage_memory.view(memory.shape[0], base_size // 2**i, base_size // 2**i, memory.shape[-1]).permute(0, 3, 1, 2)
                resized_memory = F.interpolate(stage_memory, size=(base_size, base_size), mode='bilinear', align_corners=False).unsqueeze(0)
                resized_memory = torch.norm(resized_memory, p=2, dim=2, keepdim=False)
                max_val = resized_memory.max()
                resized_memory = resized_memory / max_val
                # 存入d_score
                d_score.append(resized_memory)
            avg_score = torch.cat(d_score, dim=0)
            avg_score = torch.mean(avg_score, dim=0).squeeze()
            这是我的计算discriminability score的代码

**Comment by JiaPiangLi** [comment]
> 您好，很感谢您的工作，我使用了您的可视化代码，但是我有一个问题关于multi-scale features，据我所知，deformable detr，在encoder的操作之前，会将backbone出来的特征将H,W进行合并，然后将4维的多维特征进行铺平，然后再进行后续的操作得到feature，您这里的multi-scale feature的维度是LxBxCxHxW，请问您是将后续得到的的feature进行逆操作得到的吗，还是说我的理解有问题，期待您的回复。

你有尝试在别的网络使用Discriminability Scorej进行可视化吗，我在DINO中针对Encoder输出特征按论文步骤进行可视化发现，可视化结果与论文相反，目标物体边缘附近得分较低，背景较高

**Comment by gaowenjie-star** [comment]
> > 您好，很感谢您的工作，我使用了您的可视化代码，但是我有一个问题关于multi-scale features，据我所知，deformable detr，在encoder的操作之前，会将backbone出来的特征将H,W进行合并，然后将4维的多维特征进行铺平，然后再进行后续的操作得到feature，您这里的multi-scale feature的维度是LxBxCxHxW，请问您是将后续得到的的feature进行逆操作得到的吗，还是说我的理解有问题，期待您的回复。
> 
> 你有尝试在别的网络使用Discriminability Scorej进行可视化吗，我在DINO中针对Encoder输出特征按论文步骤进行可视化发现，可视化结果与论文相反，目标物体边缘附近得分较低，背景较高

我就是使用dino进行可视化的，我就是直接将encoder输出的特征直接用作者的可视化代码操作，不需要别的操作，就得到了正常的可视化图，注意我是在执行测试的时候才成功的，训练的时候不行
